README.md
# RESTArt-CrossDomain

A RESTArt extension for cross-domain access.


## Installation

Install `RESTArt-CrossDomain` with `pip`:

    $ pip install RESTArt-CrossDomain

Install development version from `GitHub`:

    $ pip install -e git+https://github.com/RussellLuo/restart-crossdomain.git#egg=restart-crossdomain


## Quickstart

### 1. Create the middleware class

Use the `CORSMiddleware` class:

```python
    from restart.ext.crossdomain.cors import CORSMiddleware
```

which has the default configurations as follows:

```python
    # The configuration that determines the `Access-Control-Allow-Origin` header
    cors_allow_origin = '*'  # any domain

    # The configuration that determines the `Access-Control-Allow-Credentials` header
    cors_allow_credentials = False

    # The configuration that determines the `Access-Control-Allow-Methods` header
    cors_allow_methods = ('GET', 'POST', 'PUT', 'PATCH', 'DELETE')

    # The configuration that determines the `Access-Control-Allow-Headers` header
    cors_allow_headers = ()  # any headers

    # The configuration that determines the `Access-Control-Max-Age` header
    cors_max_age = 864000  # 10 days
```

Or create a new subclass with some customized configurations:

```python
    # my_middlewares.py

    from restart.ext.crossdomain.cors import CORSMiddleware

    class CustomizedCORSMiddleware(CORSMiddleware):
        cors_allow_origin = 'http://example.com'
        cors_allow_credentials = True
```

### 2. Use the middleware class

Set the above middleware class as a [global middleware][1]:

```python
    MIDDLEWARE_CLASSES = (
        'restart.ext.crossdomain.cors.CORSMiddleware',
        # Or
        # 'my_middlewares.CustomizedCORSMiddleware',
    )
```

Or set it as a [resource-level middleware][2]:

```python
    from restart.api import RESTArt
    from restart.resource import Resource
    from restart.ext.crossdomain.cors import CORSMiddleware
    # Or
    # from my_middlewares import CustomizedCORSMiddleware as CORSMiddleware

    api = RESTArt()

    @api.route(methods=['GET'])
    class Demo(Resource):
        name = 'demo'

        middleware_classes = (CORSMiddleware,)

        def read(self, request):
            return 'this is a demo'
```


[1]: http://restart.readthedocs.org/en/latest/middleware.html#global-middlewares
[2]: http://restart.readthedocs.org/en/latest/middleware.html#resource-level-middlewares


CHANGELOG.md
# RESTArt-CrossDomain Changelog

Here you can see the full list of changes between each RESTArt-CrossDomain release.


## Version 0.1.2

Released on Jan 29th 2016.

- Handle requests without the `Origin` header correctly
- Add tests
- Update README


## Version 0.1.0

Released on Jan 24th 2016.

The first release.


tests/test_cors.py
from __future__ import absolute_import

from restart.config import config
from restart.resource import Resource
from restart.response import Response
from restart.testing import RequestFactory
from restart.ext.crossdomain.cors import CORSMiddleware


factory = RequestFactory()


class Echo(Resource):
    name = 'echo'

    middleware_classes = (CORSMiddleware,)

    def read(self, request):
        return request.data


class CustomizedCORSMiddleware(CORSMiddleware):
    cors_allow_origin = 'http://localhost'
    cors_allow_credentials = True


class AnotherEcho(Echo):
    middleware_classes = (CustomizedCORSMiddleware,)


class TestCORSMiddleware(object):

    def make_resource(self, resource_class=Echo, action_map=config.ACTION_MAP):
        return resource_class(action_map)

    def test_preflight_request_without_request_headers(self):
        request = factory.options(
            '/',
            headers=[('Origin', 'http://localhost'),
                     ('Access-Control-Request-Method', 'GET')]
        )
        resource = self.make_resource()
        response = resource.dispatch_request(request)

        assert isinstance(response, Response)
        assert response.data == '""'
        assert response.status_code == 200
        assert response.headers['Access-Control-Allow-Origin'] == '*'
        assert (response.headers['Access-Control-Allow-Methods'] ==
                'GET, POST, PUT, PATCH, DELETE')
        assert response.headers['Access-Control-Max-Age'] == '864000'
        assert 'Access-Control-Allow-Headers' not in response.headers
        assert 'Access-Control-Allow-Credentials' not in response.headers
        assert 'Vary' not in response.headers

    def test_preflight_request_with_request_headers(self):
        request = factory.options(
            '/',
            headers=[('Origin', 'http://localhost'),
                     ('Access-Control-Request-Method', 'GET'),
                     ('Access-Control-Request-Headers', 'X-CORS')]
        )
        resource = self.make_resource()
        response = resource.dispatch_request(request)

        assert isinstance(response, Response)
        assert response.data == '""'
        assert response.status_code == 200
        assert response.headers['Access-Control-Allow-Origin'] == '*'
        assert (response.headers['Access-Control-Allow-Methods'] ==
                'GET, POST, PUT, PATCH, DELETE')
        assert response.headers['Access-Control-Max-Age'] == '864000'
        assert response.headers['Access-Control-Allow-Headers'] == 'X-CORS'
        assert 'Access-Control-Allow-Credentials' not in response.headers
        assert 'Vary' not in response.headers

    def test_preflight_request_when_allow_credentials(self):
        request = factory.options(
            '/',
            headers=[('Origin', 'http://localhost'),
                     ('Access-Control-Request-Method', 'GET')]
        )
        resource = self.make_resource(resource_class=AnotherEcho)
        response = resource.dispatch_request(request)

        assert isinstance(response, Response)
        assert response.data == '""'
        assert response.status_code == 200
        assert (response.headers['Access-Control-Allow-Origin'] ==
                'http://localhost')
        assert (response.headers['Access-Control-Allow-Methods'] ==
                'GET, POST, PUT, PATCH, DELETE')
        assert response.headers['Access-Control-Max-Age'] == '864000'
        assert response.headers['Access-Control-Allow-Credentials'] == 'true'
        assert response.headers['Vary'] == 'Origin'

    def test_actual_request(self):
        data = {'hello': 'world'}
        request = factory.get('/', data=data)
        resource = self.make_resource()
        response = resource.dispatch_request(request)

        assert isinstance(response, Response)
        assert response.data == '{"hello": "world"}'
        assert response.status_code == 200
        assert response.headers['Access-Control-Allow-Origin'] == '*'
        assert 'Access-Control-Allow-Credentials' not in response.headers
        assert 'Vary' not in response.headers

    def test_actual_request_when_allow_credentials(self):
        data = {'hello': 'world'}
        request = factory.get('/', data=data)
        resource = self.make_resource(resource_class=AnotherEcho)
        response = resource.dispatch_request(request)

        assert isinstance(response, Response)
        assert response.data == '{"hello": "world"}'
        assert response.status_code == 200
        assert (response.headers['Access-Control-Allow-Origin'] ==
                'http://localhost')
        assert response.headers['Access-Control-Allow-Credentials'] == 'true'
        assert response.headers['Vary'] == 'Origin'


restart_crossdomain/cors.py
from __future__ import absolute_import


class CORSMiddleware(object):
    """The middleware used for CORS (Cross-Origin Resource Sharing).

    The CORS behaviors are implemented according to the guidelines from
    https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS
    """

    #: Access control options
    cors_allow_origin = '*'  # any domain
    cors_allow_credentials = False
    cors_allow_methods = ('GET', 'POST', 'PUT', 'PATCH', 'DELETE')
    cors_allow_headers = ()  # any headers
    cors_max_age = 864000  # 10 days

    def make_preflight_headers(self, request_headers, allow_origin=None,
                               allow_credentials=None, allow_methods=None,
                               allow_headers=None, max_age=None):
        """Make reasonable CORS response headers for preflight requests.
        """
        if allow_origin is None:
            allow_origin = self.cors_allow_origin
        if allow_credentials is None:
            allow_credentials = self.cors_allow_credentials
        if allow_methods is None:
            allow_methods = self.cors_allow_methods
        if allow_headers is None:
            allow_headers = self.cors_allow_headers
            if not allow_headers and request_headers is not None:
                allow_headers = request_headers.split(', ')
        if max_age is None:
            max_age = self.cors_max_age

        headers = {
            'Access-Control-Allow-Origin': allow_origin,
            'Access-Control-Allow-Methods': ', '.join(allow_methods),
            'Access-Control-Max-Age': '%d' % max_age,
        }
        if allow_headers:
            headers.update({
                'Access-Control-Allow-Headers': ', '.join(allow_headers)
            })
        if allow_credentials:
            headers.update({'Access-Control-Allow-Credentials': 'true'})
        if allow_origin != '*':
            headers.update({'Vary': 'Origin'})

        return headers

    def make_actual_headers(self, request_origin, allow_origin=None,
                            allow_credentials=None):
        """Make reasonable CORS response headers for actual requests.
        """
        if allow_origin is None:
            allow_origin = self.cors_allow_origin
        # If the Access-Control-Allow-Origin header is specified as
        # a wildcard, only allow the origin of the current request
        if allow_origin == '*':
            allow_origin = request_origin
        if allow_credentials is None:
            allow_credentials = self.cors_allow_credentials

        headers = {
            'Access-Control-Allow-Origin': allow_origin,
        }
        if allow_credentials:
            headers.update({'Access-Control-Allow-Credentials': 'true'})
        if allow_origin != '*':
            headers.update({'Vary': 'Origin'})

        return headers

    def is_preflight_request(self, request):
        """Judge if the `request` object is a preflight request."""
        return (request.method == 'OPTIONS' and
                'Origin' in request.headers and
                'Access-Control-Request-Method' in request.headers)

    def process_request(self, request):
        """Handle the preflight request correctly.

        :param request: the request object.
        """
        if self.is_preflight_request(request):
            request_headers = request.headers.get(
                'Access-Control-Request-Headers'
            )
            headers = self.make_preflight_headers(request_headers)
            return '', 200, headers

    def process_response(self, request, response):
        """Add appropriate response headers for the actual request.

        :param request: the request object.
        :param response: the response object.
        """
        if not self.is_preflight_request(request):
            request_origin = request.headers['Origin']
            headers = self.make_actual_headers(request_origin)
            response.headers.update(headers)
        return response
